packages/solidity-contracts/contracts/fuelchain/FuelMessagePortal/v3/FuelMessagePortalV3.sol

```diff
@@ -9,13 +9,43 @@ contract FuelMessagePortalV3 is FuelMessagePortalV2 {
     using FuelBlockHeaderLiteLib for FuelBlockHeaderLite;
 
     error MessageBlacklisted();
-    error WithdrawalsPaused();
     error MessageRelayFailed();
+    error RateLimitExceeded();
+    error WithdrawalsPaused();
+
+    /// @dev The rate limit setter role
+    bytes32 public constant SET_RATE_LIMITER_ROLE = keccak256("SET_RATE_LIMITER_ROLE");
+
+    /// @notice Duration after which rate limit resets.
+    uint256 public immutable rateLimitDuration;
+
+    /// @notice Amounts already withdrawn this period.
+    uint256 public currentPeriodAmount;
 
+    /// @notice Flag to indicate whether withdrawals are paused or not.
     bool public withdrawalsPaused;
+
+    /// @notice The time at which the current period ends at.
+    uint256 public currentPeriodEnd;
+
+    /// @notice The eth withdrawal limit amount.
+    uint256 public limitAmount;
+
     mapping(bytes32 => bool) public messageIsBlacklisted;
 
-    constructor(uint256 _depositLimitGlobal) FuelMessagePortalV2(_depositLimitGlobal) {}
+    constructor(uint256 _depositLimitGlobal, uint256 _rateLimitDuration) FuelMessagePortalV2(_depositLimitGlobal) {
+        rateLimitDuration = _rateLimitDuration;
+    }
+
+    function initializerV3(FuelChainState fuelChainState, uint256 _limitAmount) public initializer {
+        super.initialize(fuelChainState);
+        _setInitParams(_limitAmount);
```

Comment:
Does UUPS allow to call reinitializeV3 here? As things stand, I think `reinitializeV3` can be called again after a normal deployment.

Add a unit test to make sure that neither `initializerV3` nor `reinitializeV3` can be called after deploying the proxy with the plugin

Reply:
no we can't because then the `reinitialize` modifier get's invoked and the tx get's reverts hence added this init logic in a internal

Reply:
I see.

Research if we can use `reinitializer(3)` instead of `initializer` modifier. We need to add a unit test to ensure that these methods (`initializeV3`, `reinitializeV3`)  can only be called once

Reply:
hm but why ? because when we deploy it from scratch then we won't need `reinitializeV3 ` and I think we have to use `initializer` for the default version

Reply:
check https://docs.openzeppelin.com/contracts/4.x/api/proxy#Initializable